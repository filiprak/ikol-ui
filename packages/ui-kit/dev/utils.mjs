import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
