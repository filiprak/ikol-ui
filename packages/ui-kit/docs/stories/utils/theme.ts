import { useTheme } from '@ui/composables';
import { createIkolUI } from '@ui/index';
import { App } from 'vue';
import { StoryContext, VueRenderer } from '@storybook/vue3';

export function setupTheme(context: any) {
    const theme = useTheme();

    theme.is_dark.value = context?.globals?.theme === 'dark';

    return theme;
}

export function setupApp(app: App, context?: StoryContext<VueRenderer>) {
    app.use(createIkolUI());
}
