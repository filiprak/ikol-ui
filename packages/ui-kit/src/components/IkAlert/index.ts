import IkAlert from './IkAlert.vue';
export interface IkAlert extends InstanceType<typeof IkAlert> { };
export { IkAlert };
