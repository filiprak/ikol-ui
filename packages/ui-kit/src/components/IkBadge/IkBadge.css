.ik-badge {
    position: relative;
    display: inline-block;
}

.ik-badge--outer {
    display: flex;
    flex-direction: row;
    align-items: center;
}

.ik-badge__dot {
    display: flex;
    align-items: center;
    justify-content: center;
    position: absolute;
    bottom: 0;
    right: 0;
    border-radius: 50%;
    font-size: 11px;
}

.ik-badge--outer .ik-badge__dot {
    position: static;
    margin-left: 5px;
    flex: 0 0 auto;
}

.ik-badge--outer .ik-badge__content {
    flex: 1 1 auto;
    overflow: hidden;
    text-overflow: ellipsis;
}

.ik-badge--top .ik-badge__dot {
    top: 0;
    bottom: auto;
}

.ik-badge--left .ik-badge__dot {
    left: 0;
    right: auto;
}

.ik-badge--success .ik-badge__dot {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-badge--error .ik-badge__dot {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-badge--accent .ik-badge__dot {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-badge--primary .ik-badge__dot {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-badge__dot,
.ik-badge--default .ik-badge__dot {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-badge--transparent .ik-badge__dot {
    background-color: transparent;
    color: inherit;
}