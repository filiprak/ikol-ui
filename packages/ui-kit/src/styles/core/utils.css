/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
