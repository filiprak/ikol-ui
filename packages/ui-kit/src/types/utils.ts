import type { ExtractPropTypes } from 'vue';

/* eslint-disable @typescript-eslint/naming-convention */
export type IkExtractProps<T> =
    T extends new (...angs: any) => { $props: infer P; } ? NonNullable<P> : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: infer P, ...args: any) => any ? P : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkExtractPropTypes<T> = Prettify<Readonly<ExtractPropTypes<T>>>;

export type IkExtractExposed<T> =
    T extends new (...angs: any) => infer E ? E : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: any, ctx: any, expose: (exposed: infer E) => any, ...args: any) => any ? NonNullable<E> : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkKeysByType<T, U> = keyof {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

export enum IkUIDesignColor {
    primary = 'primary',
    success = 'success',
    error = 'error',
    accent = 'accent',
    default = 'default',
}

export enum IkUIElemSize {
    xs = 'xs',
    sm = 'sm',
    md = 'md',
    lg = 'lg',
}

export enum IkUIVariant {
    flat = 'flat',
    filled = 'filled',
    outline = 'outline',
    pill = 'pill',
}

export enum IkPosition {
    top = 'top',
    right = 'right',
    bottom = 'bottom',
    left = 'left',
    horizontal = 'horizontal',
    vertical = 'vertical',
}

export enum IkAlignment {
    start = 'start',
    center = 'center',
    end = 'end',
}

export type EnumT = Record<string, string> | Record<string | number, string | number>;
export type ObjValueT<T> = T[keyof T];
export type DefinedKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T];
export type OmitUndefined<T> = { [K in DefinedKeys<T>]: T[K] };
export type NotArray<T> = T extends any[] ? never : T; // eslint-disable-line @typescript-eslint/no-explicit-any
export type DefinedComponent = new (...args: any[]) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
