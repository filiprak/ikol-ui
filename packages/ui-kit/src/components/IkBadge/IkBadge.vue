<script lang="ts">
import '@ui/styles';
import './IkBadge.css';
import { computed, h, defineComponent, type PropType } from 'vue';
import type { IkUIDesignColor } from '@ui/types/utils';

export default defineComponent({
    props: {
        design: {
            type: String as PropType<`${IkUIDesignColor}`>,
        },
        top: {
            type: Boolean,
            default: false,
        },
        left: {
            type: Boolean,
            default: false,
        },
        size: {
            type: Number,
            default: 12,
        },
        number: {
            type: [Number, String],
        },
        show: {
            type: Boolean,
            default: true,
        },
        show_zero: {
            type: Boolean,
            default: false,
        },
        outer: {
            type: Boolean,
            default: false,
        },
    },
    setup(props, { slots }) {
        const classes = computed(() => {
            return {
                'ik-badge': true,
                'ik-badge--outer': props.outer,
                'ik-badge--top': props.top && !props.outer,
                'ik-badge--left': props.left && !props.outer,
                ['ik-badge--' + props.design]: !!props.design,
            };
        });

        let show: boolean = props.show;

        let mask: string | null = null;

        if (!props.outer) {
            const r = Math.round(props.size / 2);
            const mr = (r + 2);

            const mx = props.left ? (r + 'px') : ('calc(100% - ' + r + 'px)');
            const my = props.top ? (r + 'px') : ('calc(100% - ' + r + 'px)');

            mask = `radial-gradient(circle ${mr}px at ${mx} ${my}, transparent 95%, #fff 100%)`;
        }

        let dot_text = '';

        if (!Number.isNaN(Number(props.number))) {
            if (Number(props.number) == 0 && !props.show_zero) {
                show = false;
            }
            if (Number(props.number) < 10) {
                dot_text = String(props.number);
            } else {
                dot_text = '9+';
            }
        }

        return () => h('div', {
            class: classes.value,
        }, [
            h('div', {
                class: 'ik-badge__content',
                style: show && mask
                    ? {
                        mask,
                        '-webkit-mask': mask,
                    }
                    : null,
            }, slots.default?.()),
            show
                ? h('div', {
                    class: 'ik-badge__dot',
                    style: { width: props.size + 'px', height: props.size + 'px' },
                }, slots.content?.() || dot_text)
                : null,
        ]);
    },
});
</script>
