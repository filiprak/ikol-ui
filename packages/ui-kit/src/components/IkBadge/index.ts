import IkBadge from './IkBadge.vue';
export interface IkBadge extends InstanceType<typeof IkBadge> { };
export { IkBadge };
