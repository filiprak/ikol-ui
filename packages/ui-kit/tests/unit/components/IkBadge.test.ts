import { renderDeep } from '@ui/tests/tools/utils';
import { IkBadge } from '@ui/components/IkBadge';

describe('IkBadge', () => {
    const getInstance = (options = {}) => renderDeep(IkBadge, options);

    it('renders with correct default state', () => {
        const wrapper = getInstance();
        expect(wrapper.classes()).toContain('ik-badge');
        expect(wrapper.element.tagName).toBe('DIV');

        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with custom design', () => {
        const wrapper = getInstance({
            props: { design: 'error' },
        });
        expect(wrapper.classes()).toContain('ik-badge--error');
    });

    it('renders with custom size', () => {
        const wrapper = getInstance({
            props: { size: 24 },
        });

        const dot_styles = wrapper.find('.ik-badge__dot').attributes('style');
        const content_styles = wrapper.find('.ik-badge__content').attributes('style');

        expect(dot_styles).toContain('width: 24px');
        expect(dot_styles).toContain('height: 24px');
        expect(content_styles).toContain('circle 14px at calc(100% - 12px) calc(100% - 12px), ');
    });

    it('renders left', () => {
        const wrapper = getInstance({
            props: { size: 24, left: true },
        });

        const dot_styles = wrapper.find('.ik-badge__dot').attributes('style');
        const content_styles = wrapper.find('.ik-badge__content').attributes('style');

        expect(dot_styles).toContain('width: 24px');
        expect(dot_styles).toContain('height: 24px');
        expect(content_styles).toContain('circle 14px at 12px calc(100% - 12px), ');
    });

    it('renders top', () => {
        const wrapper = getInstance({
            props: { size: 24, top: true },
        });

        const dot_styles = wrapper.find('.ik-badge__dot').attributes('style');
        const content_styles = wrapper.find('.ik-badge__content').attributes('style');

        expect(dot_styles).toContain('width: 24px');
        expect(dot_styles).toContain('height: 24px');
        expect(content_styles).toContain('circle 14px at calc(100% - 12px) 12px, ');
    });

    it('renders with default slot', () => {
        const wrapper = getInstance({
            slots: {
                default: '<div>This is badge content...</div>',
            },
        });

        expect(wrapper.find('.ik-badge__content').text())
            .toContain('This is badge content...');
    });

    it('renders with dot content slot', () => {
        const wrapper = getInstance({
            slots: {
                content: '<div>This is dot content</div>',
            },
        });

        expect(wrapper.find('.ik-badge__dot').text())
            .toContain('This is dot content');
    });

    it('renders with dot hidden', () => {
        const wrapper = getInstance({
            props: { show: false },
        });

        expect(wrapper.find('.ik-badge__content').attributes('style')).toBeFalsy();
        expect(wrapper.find('.ik-badge__dot').exists()).toBeFalsy();
    });

    it('renders with number values in dot content', () => {
        let wrapper;

        wrapper = getInstance({ props: { number: 0 } });
        expect(wrapper.find('.ik-badge__dot').exists()).toBeFalsy();

        wrapper = getInstance({ props: { number: 1 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('1');

        wrapper = getInstance({ props: { number: -1 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('-1');

        wrapper = getInstance({ props: { number: -10 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('-10');

        wrapper = getInstance({ props: { number: 9 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('9');

        wrapper = getInstance({ props: { number: 10 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('9+');

        wrapper = getInstance({ props: { number: 235 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('9+');
    });

    it('renders with number values in dot content including zero', () => {
        let wrapper;

        wrapper = getInstance({ props: { show_zero: true, number: 0 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('0');

        wrapper = getInstance({ props: { show_zero: true, number: 1 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('1');

        wrapper = getInstance({ props: { show_zero: true, number: -1 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('-1');

        wrapper = getInstance({ props: { show_zero: true, number: -10 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('-10');

        wrapper = getInstance({ props: { show_zero: true, number: 9 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('9');

        wrapper = getInstance({ props: { show_zero: true, number: 10 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('9+');

        wrapper = getInstance({ props: { show_zero: true, number: 235 } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('9+');

        wrapper = getInstance({ props: { show: false, show_zero: true, number: 235 } });
        expect(wrapper.find('.ik-badge__dot').exists()).toBeFalsy();
    });

    it('renders with numeric string values in dot content', () => {
        let wrapper;

        wrapper = getInstance({ props: { show_zero: true, number: '0' } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('0');

        wrapper = getInstance({ props: { show_zero: false, number: '0' } });
        expect(wrapper.find('.ik-badge__dot').exists()).toBeFalsy();

        wrapper = getInstance({ props: { show_zero: true, number: 'abcd' } });
        expect(wrapper.find('.ik-badge__dot').text()).toBe('');
    });

    it('renders outer', () => {
        const wrapper = getInstance({
            props: {
                outer: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-badge--outer');
        expect(wrapper.element).toMatchSnapshot();
    });
});
