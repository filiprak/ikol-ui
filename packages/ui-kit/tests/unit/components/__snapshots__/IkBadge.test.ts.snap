// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkBadge renders outer 1`] = `
<div
  class="ik-badge ik-badge--outer"
>
  <div
    class="ik-badge__content"
  />
  <div
    class="ik-badge__dot"
    style="width: 12px; height: 12px;"
  />
</div>
`;

exports[`IkBadge renders with correct default state 1`] = `
<div
  class="ik-badge"
>
  <div
    class="ik-badge__content"
    style="mask: radial-gradient(circle 8px at calc(100% - 6px) calc(100% - 6px), transparent 95%, #fff 100%);"
  />
  <div
    class="ik-badge__dot"
    style="width: 12px; height: 12px;"
  />
</div>
`;
